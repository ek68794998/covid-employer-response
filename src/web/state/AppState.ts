import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}