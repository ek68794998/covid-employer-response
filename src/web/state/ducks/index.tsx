import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});