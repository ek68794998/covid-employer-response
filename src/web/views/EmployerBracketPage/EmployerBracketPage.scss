@import "../Colors.scss";
@import "../Dimensions.scss";

$division-padding: 16px;

main#employer-bracket {
	margin: 0 auto;
	width: 100%;

	.EmployerBracketPage__Container {
		padding: 0 $page-padding;
	}

	.EmployerBracketPage__Content {
		display: flex;
		margin: 0 auto;
		max-width: $body-max-width;
		padding-bottom: $division-padding;
		padding-top: $division-padding;

		.EmployerBracketPage__ContentSpacer {
			width: 48px;
		}

		.EmployerBracketList__Container {
			flex: 1;
		}
	}
}

@media (max-width: $body-narrow-width) {
	main#employer-bracket {
		.EmployerBracketPage__Container {
			padding: 0 $page-padding-narrow;
		}
	}
}
