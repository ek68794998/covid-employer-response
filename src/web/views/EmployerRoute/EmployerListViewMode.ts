// For maintainability of union types, do not require a typedef.
// eslint-disable-next-line
export const EmployerListViewModeValues = [
	"cards",
	"rows",
] as const;

export type EmployerListViewMode = typeof EmployerListViewModeValues[number];
