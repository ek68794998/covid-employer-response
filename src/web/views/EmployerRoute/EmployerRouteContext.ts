/* eslint-disable unused-imports/no-unused-vars-ts */
/* Methods declared in the interface must be used in the context default value, but don't use their arguments. */

import React from "react";

import { EmployerListSearchFilter } from "../EmployerListSearch/EmployerListSearchFilter";

type ContextDataSetter<T> = (value: T) => void;

const getDefaultContextDataSetter = <T>(): ContextDataSetter<T> =>
	(value: T): void => (((): void => { /* Noop. */ }) as any)(value);

export const DefaultContextData: EmployerRouteContextData = {
	pageIndex: 1,
	searchFilters: new EmployerListSearchFilter(),
	setPageIndex: getDefaultContextDataSetter<number>(),
	setSearchFilters: getDefaultContextDataSetter<EmployerListSearchFilter>(),
};

export interface EmployerRouteContextData {
	pageIndex: number;

	searchFilters: EmployerListSearchFilter;

	setPageIndex: ContextDataSetter<number>;

	setSearchFilters: ContextDataSetter<EmployerListSearchFilter>;
}

export const EmployerRouteContext: React.Context<EmployerRouteContextData> =
	React.createContext(DefaultContextData);
