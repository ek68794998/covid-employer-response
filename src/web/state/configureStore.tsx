import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;