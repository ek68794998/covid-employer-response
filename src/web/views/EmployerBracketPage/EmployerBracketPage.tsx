import React from "react";
import { useSelector } from "react-redux";

import { EmployerRecordMetadata } from "../../../common/EmployerRecordMetadata";
import { LocalizedStrings } from "../../../common/LocalizedStrings";
import { getEmployersList } from "../../state/ducks/employers/selectors";
import { getStrings } from "../../state/ducks/localization/selectors";

import AppHelmet from "../AppHelmet/AppHelmet";
import EmployerBracketList from "../EmployerBracketList/EmployerBracketList";
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator";

import "./EmployerBracketPage.scss";

interface Props {
	mode: "recent" | "ranking";
}

const defaultDate: string = "2000-01-01T00:00:00Z";

const EmployerBracketPage: React.FC<Props> = (props: Props): React.ReactElement | null => {
	const strings: LocalizedStrings = useSelector(getStrings);
	const employersList: EmployerRecordMetadata[] | undefined = useSelector(getEmployersList);

	const { mode } = props;

	if (!employersList) {
		return <LoadingIndicator />;
	}

	const sortedEmployers: EmployerRecordMetadata[] =
		employersList
			.concat() // Copy the array.
			.sort((a: EmployerRecordMetadata, b: EmployerRecordMetadata) => {
				if (mode === "ranking") {
					return a.score - b.score;
				}

				if (mode === "recent") {
					return (
						new Date(a.lastUpdated || defaultDate).getTime()
						- new Date(b.lastUpdated || defaultDate).getTime()
					);
				}

				return 0;
			});

	const numberToShow: number = 10;

	const bottomEmployers: EmployerRecordMetadata[] =
		sortedEmployers.slice(0, numberToShow);

	const topEmployers: EmployerRecordMetadata[] =
		sortedEmployers.slice(-numberToShow).reverse();

	if (mode === "ranking") {
		return (
			<main id="employer-bracket">
				<AppHelmet
					title={strings.employerListRanking}
				/>
				<div className="EmployerBracketPage__Container">
					<div className="EmployerBracketPage__Content">
						<EmployerBracketList
							employers={topEmployers}
							metric="score"
							title={strings.employerListRankingBestTitle}
						/>
						<div className="EmployerBracketPage__ContentSpacer" />
						<EmployerBracketList
							employers={bottomEmployers}
							metric="score"
							title={strings.employerListRankingWorstTitle}
						/>
					</div>
				</div>
			</main>
		);
	}

	if (mode === "recent") {
		return (
			<main id="employer-bracket">
				<AppHelmet
					title={strings.employerListRecentTitle}
				/>
				<div className="EmployerBracketPage__Container">
					<div className="EmployerBracketPage__Content">
						<EmployerBracketList
							employers={topEmployers}
							metric="lastUpdated"
							title={strings.employerListRecentTitle}
						/>
					</div>
				</div>
			</main>
		);
	}

	return null;
};

export default EmployerBracketPage;
