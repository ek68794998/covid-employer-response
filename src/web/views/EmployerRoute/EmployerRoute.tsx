import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";

import EmployerBracketList from "../EmployerBracketList/EmployerBracketList";
import EmployerListPage from "../EmployerListPage/EmployerListPage";
import EmployerPage from "../EmployerPage/EmployerPage";

import { EmployerListViewMode, EmployerListViewModeValues } from "./EmployerListViewMode";
import { DefaultContextData, EmployerRouteContext, EmployerRouteContextData } from "./EmployerRouteContext";

export const RankingsRoute: string = "ranking";
export const RecentsRoute: string = "recent";

interface Params {
	id?: string;
}

type Props = RouteComponentProps<Params>;

const canUseSessionStorage: boolean = typeof sessionStorage !== "undefined";

const EmployerRoute: React.FC<Props> = (props: Props): React.ReactElement => {
	const [ listViewMode, setListViewMode ] = useState(DefaultContextData.listViewMode);
	const [ pageIndex, setPageIndex ] = useState(DefaultContextData.pageIndex);
	const [ searchFilters, setSearchFilters ] = useState(DefaultContextData.searchFilters);

	const contextData: EmployerRouteContextData = {
		listViewMode,
		pageIndex,
		searchFilters,
		setListViewMode,
		setPageIndex,
		setSearchFilters,
	};

	if (canUseSessionStorage) {
		useEffect(
			(): void => {
				const sessionListViewMode: string | null =
					canUseSessionStorage ? sessionStorage.getItem("EmployerRoute.listViewMode") : null;

				if (EmployerListViewModeValues.indexOf(sessionListViewMode as any) >= 0) {
					setListViewMode(sessionListViewMode as EmployerListViewMode);
				}
			},
			[]);

		useEffect(
			(): void => {
				sessionStorage.setItem("EmployerRoute.listViewMode", listViewMode);
			},
			[ listViewMode ]);
	}

	const employerId: string | undefined = props.match.params.id;

	let employerRoutedComponent: JSX.Element;

	if (employerId) {
		if (employerId === RecentsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="recent" />;
		} else if (employerId === RankingsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="ranking" />;
		} else {
			employerRoutedComponent = <EmployerPage employerId={employerId} />;
		}
	} else {
		employerRoutedComponent = <EmployerListPage />;
	}

	return (
		<EmployerRouteContext.Provider value={contextData}>
			{employerRoutedComponent}
		</EmployerRouteContext.Provider>
	);
};

export default EmployerRoute;
