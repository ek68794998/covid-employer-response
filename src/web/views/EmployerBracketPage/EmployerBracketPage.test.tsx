import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { getPlocStringsAsync, mockComponent } from "../../../__tests__/TestUtils";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import { EmployerRouteContext, DefaultContextData } from "../EmployerRoute/EmployerRouteContext";

import EmployerBracketPage from "./EmployerBracketPage";

jest.mock(
	"../EmployerBracketList/EmployerBracketList",
	() => mockComponent("EmployerBracketList"));

jest.mock(
	"../LoadingIndicator/LoadingIndicator",
	() => mockComponent("LoadingIndicator"));

describe("<EmployerBracketPage />", () => {
	test("renders without exploding", async () => {
		const store: Store<AppState, AnyAction> = configureStore({ strings: await getPlocStringsAsync() });

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerRouteContext.Provider value={DefaultContextData}>
							<EmployerBracketPage mode="ranking" />
						</EmployerRouteContext.Provider>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
