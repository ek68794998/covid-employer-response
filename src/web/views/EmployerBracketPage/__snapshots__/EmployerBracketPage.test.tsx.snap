// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerBracketPage /> renders without exploding 1`] = `"[component: LoadingIndicator /]"`;
